Amazon.QLDB.Driver/session/AsyncSessionPool.cs

```diff
@@ -0,0 +1,175 @@
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+namespace Amazon.QLDB.Driver
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Extensions.Logging;
+
+    /// <summary>
+    /// The implementation of the session pool.
+    /// </summary>
+    internal class AsyncSessionPool : BaseSessionPool
+    {
+        private readonly BlockingCollection<AsyncQldbSession> sessionPool;
+        private new readonly SemaphoreSlim poolPermits;
+        private readonly Func<CancellationToken, Task<Session>> sessionCreator;
+        private readonly IAsyncRetryHandler retryHandler;
+        private readonly ILogger logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncSessionPool"/> class.
+        /// </summary>
+        /// <param name="sessionCreator">The method to create a new underlying QLDB session. The operation can be cancelled.</param>
+        /// <param name="retryHandler">Handling the retry logic of the execute call.</param>
+        /// <param name="maxConcurrentTransactions">The maximum number of sessions that can be created from the pool at any one time.</param>
+        /// <param name="logger">Logger to be used by this.</param>
+        public AsyncSessionPool(
+            Func<CancellationToken, Task<Session>> sessionCreator,
+            IAsyncRetryHandler retryHandler,
+            int maxConcurrentTransactions,
+            ILogger logger)
+        {
+            this.sessionPool = new BlockingCollection<AsyncQldbSession>(maxConcurrentTransactions);
+            this.poolPermits = new SemaphoreSlim(maxConcurrentTransactions, maxConcurrentTransactions);
+            this.sessionCreator = sessionCreator;
+            this.retryHandler = retryHandler;
+            this.logger = logger;
+        }
+
+        /// <summary>
+        /// Execute a function asynchronously in session pool.
+        /// </summary>
+        /// <typeparam name="T">The return type of the function.</typeparam>
+        ///
+        /// <param name="func">The function to be executed in the session pool. The operation can be cancelled.</param>
+        /// <param name="retryPolicy">The policy on retry.</param>
+        /// <param name="retryAction">The customer retry action. The operation can be cancelled.</param>
+        /// <param name="cancellationToken">
+        ///     A cancellation token that can be used by other objects or threads to receive notice of cancellation.
+        /// </param>
+        ///
+        /// <returns>The result from the function.</returns>
+        public async Task<T> Execute<T>(Func<AsyncTransactionExecutor, Task<T>> func, RetryPolicy retryPolicy, Func<int, Task> retryAction, CancellationToken cancellationToken = default)
```

Comment:
retryAction is deprecated so we should not have it in the async impl.

Reply:
I removed retryAction. Should I put it back?